File: CredUICore/CUIField.cpp
//
//  CUIField.cpp
//  CredUI
//
//  Created by Luke Howard on 28/12/2013.
//  Copyright (c) 2013 PADL Software Pty Ltd. All rights reserved.
//

#include "CredUICore_Private.h"

static CFTypeID __CUIFieldTypeID = _kCFRuntimeNotATypeID;
static CFStringRef __CUIFieldValueProperty = CFSTR("value");
static CFStringRef __CUIFieldOptionsProperty = CFSTR("options");

CF_INLINE Boolean __CFEqualNullSafe(CFTypeRef cf1, CFTypeRef cf2)
{
    if (cf1 && cf2 == NULL)
        return false;
    if (cf2 && cf1 == NULL)
        return false;
    if (cf1)
        return CFEqual(cf1, cf2);
    
    return true;
}

static CFTypeRef _CUIFieldCopy(CFAllocatorRef allocator, CFTypeRef cf)
{
    return (CFTypeRef)CUIFieldCreateCopy(allocator, (CUIFieldRef)cf);
}

static void _CUIFieldDeallocate(CFTypeRef cf)
{
    CUIFieldRef field = (CUIFieldRef)cf;
    
    if (field->_title)
        CFRelease(field->_title);
    if (field->_defaultValue)
        CFRelease(field->_defaultValue);
    if (field->_delegate)
        _Block_release(field->_delegate);
}

static Boolean _CUIFieldEqual(CFTypeRef cf1, CFTypeRef cf2)
{
    Boolean equal;
    CUIFieldRef f1 = (CUIFieldRef)cf1, f2 = (CUIFieldRef)cf2;
    
    equal = (cf1 == cf2);
    
    if (!equal) {
        equal = (f1->_class == f2->_class) &&
        __CFEqualNullSafe(f1->_title, f2->_title) &&
        __CFEqualNullSafe(f1->_defaultValue, f2->_defaultValue);
    }
    
    return equal;
    
}

static CFStringRef _CUIFieldCopyDescription(CFTypeRef cf)
{
    CUIFieldRef field = (CUIFieldRef)cf;
    CFStringRef desc;
    
    desc = CFStringCreateWithFormat(CFGetAllocator(cf), NULL,
                                    CFSTR("<CUIField %p>{class = %d, title = \"%@\", options = %08x}"),
                                    field,
                                    (int)field->_class, field->_title, (unsigned int)field->_options);
    
    return desc;
}

static const CFRuntimeClass _CUIFieldClass = {
    0,
    "CUIField",
    NULL, // init
    _CUIFieldCopy, // copy
    _CUIFieldDeallocate,
    _CUIFieldEqual,
    NULL, // _CUIFieldHash,
    NULL, // _CUICopyFormattingDesc
    _CUIFieldCopyDescription
};

CUI_EXPORT CFTypeID
CUIFieldGetTypeID(void)
{
    static dispatch_once_t onceToken;
    
    dispatch_once(&onceToken, ^{
        if (__CUIFieldTypeID == _kCFRuntimeNotATypeID) {
            __CUIFieldTypeID = _CFRuntimeRegisterClass(&_CUIFieldClass);
            _CFRuntimeBridgeClasses(__CUIFieldTypeID, "CUICFField");
        }
    });
    
    return __CUIFieldTypeID;
}

CUI_EXPORT CUIFieldRef
CUIFieldCreate(CFAllocatorRef allocator,
               CUIFieldClass fieldClass,
               CFStringRef title,
               CFTypeRef defaultValue,
               void (^fieldDidChange)(CUIFieldRef field, CFTypeRef value))
{
    CUIFieldRef f;
    
    f = (CUIFieldRef)_CFRuntimeCreateInstance(allocator, CUIFieldGetTypeID(),
                                              sizeof(struct __CUIField) - sizeof(CFRuntimeBase), NULL);
    if (f == NULL)
        return NULL;
    
    f->_class = fieldClass;
    f->_title = title ? (CFStringRef)CFRetain(title) : NULL;
    f->_defaultValue = defaultValue ? CFRetain(defaultValue) : NULL;
    f->_delegate = (void (^)(CUIFieldRef, CFTypeRef))_Block_copy(fieldDidChange);
    
    return f;
}

CUI_EXPORT CUIFieldRef
CUIFieldCreateCopy(CFAllocatorRef allocator,
                   CUIFieldRef field)
{
    CF_OBJC_FUNCDISPATCHV(__CUIFieldTypeID, CUIFieldRef, field, "copy");
    
    CUIFieldRef f = CUIFieldCreate(allocator, field->_class, field->_title, field->_defaultValue, field->_delegate);
    if (f == NULL)
        return NULL;
    
    return f;
}

CUI_EXPORT CUIFieldClass
CUIFieldGetClass(CUIFieldRef field)
{
    CF_OBJC_FUNCDISPATCHV(__CUIFieldTypeID, CUIFieldClass, field, "fieldClass");

    return field->_class;
}

CUI_EXPORT CFStringRef
CUIFieldGetTitle(CUIFieldRef field)
{
    CF_OBJC_FUNCDISPATCHV(__CUIFieldTypeID, CFStringRef, field, "title");

    return field->_title;
}

CUI_EXPORT CFTypeRef
CUIFieldGetDefaultValue(CUIFieldRef field)
{
    CF_OBJC_FUNCDISPATCHV(__CUIFieldTypeID, CFTypeRef, field, "defaultValue");

    return field->_defaultValue;
}

CUI_EXPORT void
CUIFieldSetValue(CUIFieldRef field, CFTypeRef value)
{
    CF_OBJC_FUNCDISPATCHV(__CUIFieldTypeID, void, field, "setValue:", value);

    CF_OBJC_KVO_WILLCHANGE(field, __CUIFieldValueProperty);
    if (field->_delegate)
        field->_delegate(field, value);
    CF_OBJC_KVO_DIDCHANGE(field, __CUIFieldValueProperty);
}

CUI_EXPORT void
CUIFieldSetOptions(CUIFieldRef field, CUIFieldOptions value)
{
    CF_OBJC_FUNCDISPATCHV(__CUIFieldTypeID, void, field, "setOptions:", value);
    
    CF_OBJC_KVO_WILLCHANGE(field, __CUIFieldOptionsProperty);
    field->_options = value;
    CF_OBJC_KVO_DIDCHANGE(field, __CUIFieldOptionsProperty);
}

CUI_EXPORT CUIFieldOptions
CUIFieldGetOptions(CUIFieldRef field)
{
    CF_OBJC_FUNCDISPATCHV(__CUIFieldTypeID, BOOL, field, "options");
    
    return field->_options;
}

File: CredUI/CUIField.m
//
//  CUIField.m
//  CredUI
//
//  Created by Luke Howard on 30/12/2013.
//  Copyright (c) 2013 PADL Software Pty Ltd. All rights reserved.
//

#pragma mark - CUICFField concrete class

@interface CUICFField : CUIField
@end

@implementation CUICFField

+ (id)allocWithZone:(NSZone *)zone
{
    static CUIFieldRef placeholderField;
    static dispatch_once_t onceToken;
    
    dispatch_once(&onceToken, ^{
        if (placeholderField == NULL)
            placeholderField = CUIFieldCreate(kCFAllocatorDefault, kCUIFieldClassInvalid,
                                              NULL, NULL, NULL);
    });
    
    return (id)placeholderField;
}

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-implementations"
CF_CLASSIMPLEMENTATION(CUICFField)
#pragma clang diagnostic pop

+ (BOOL)automaticallyNotifiesObserversForKey:(NSString *)theKey
{
    return NO;
}

- init
{
    [super init];
    NSAssert(0, @"CUICFField can only be instantiated through CredUICore");
    return nil;
}

- (CUIFieldClass)fieldClass
{
    return CUIFieldGetClass([self _fieldRef]);
}

- (NSString *)title
{
    return (NSString *)CUIFieldGetTitle([self _fieldRef]);
}

- (id)defaultValue
{
    return (id)CUIFieldGetDefaultValue([self _fieldRef]);
}

- (CUIFieldOptions)options
{
    return CUIFieldGetOptions([self _fieldRef]);
}

- (void)setOptions:(CUIFieldOptions)options
{
    CUIFieldSetOptions([self _fieldRef], options);
}

- (void)setValue:(id)aValue
{
    id aValueCopy = [aValue copy];
    
    CUIFieldSetValue([self _fieldRef], (CFTypeRef)aValueCopy);
    
    [aValueCopy release];
}

@end

#pragma mark - CUIField abstract class

@implementation CUIField

- (CUIFieldRef)_fieldRef
{
    return (CUIFieldRef)self;
}

- (CFTypeID)_cfTypeID
{
    return CUIFieldGetTypeID();
}

#pragma mark Default methods

- (id)valueForUndefinedKey:(NSString *)key
{
    /* The value is actually write-only, so just return a defaultValue for KVO compliance */
    if ([key isEqualToString:@"value"])
        return [self defaultValue];
    else
        return [super valueForUndefinedKey:key];
}

- (void)didSubmit:(id)sender
{
    [self setValue:@YES];
}

- (void)controlTextDidChange:(NSNotification *)notification
{
    NSTextView *textView = [notification.userInfo objectForKey:@"NSFieldEditor"];
    
    [self setValue:textView.string];
}

- (BOOL)control:(NSControl *)control textView:(NSTextView *)textView doCommandBySelector:(SEL)command
{
    if (command == @selector(insertNewline:)) {
        id identityPicker = [[(NSView *)textView window] delegate];
        
        [identityPicker willSubmitCredential:control];
        
        return YES;
    }
    
    return NO;
}

#pragma mark Primitive methods

- (void)setValue:(id)aValue
{
    NSRequestConcreteImplementation(self, _cmd, [CUIField class]);
}

- (void)setOptions:(CUIFieldOptions)options
{
    NSRequestConcreteImplementation(self, _cmd, [CUIField class]);
}

- (CUIFieldClass)fieldClass
{
    return kCUIFieldClassInvalid;
}

- (NSString *)title
{
    return nil;
}

- (id)defaultValue
{
    return nil;
}

- (CUIFieldOptions)options
{
    return kCUIFieldOptionsNone;
}

@end
